dockers/docker-sonic-telemetry/certificate_rotation_checker

```diff
@@ -0,0 +1,205 @@
+        return 2, command_stderr
+
+    return proc_instance.returncode, command_stdout.strip()
+
+
+def check_telemetry_server_running():
+    """Checkes whether telemetry server process is running.
+
+    Args:
+        None.
+
+    Returns:
+        None.
+    """
+    processes_status_cmd = "supervisorctl status"
+    retry_times = 0
+    is_running = False
+
+    while retry_times <= MAX_RETRY_TIMES:
+        retry_times += 1
+        exit_code, command_stdout = get_command_result(processes_status_cmd)
+        if exit_code != 3:
+            syslog.syslog(syslog.LOG_INFO,
+                          "Failed to get the processes running status in telemetry container and retry after 60 seconds ...")
+            time.sleep(60)
+        else:
+            for line in command_stdout.splitlines():
+                if "telemetry" in line and "RUNNING" in line:
+                    is_running = True
+                    break
+            if is_running:
+                syslog.syslog(syslog.LOG_INFO,
+                              "Telemetry server process is running after certificate and private key were rotated!")
+                break
+
+    if not is_running:
+        syslog.syslog(syslog.LOG_ERR,
+                      "Telemetry server process is not running after certificate and private key were rotated and exiting ...")
+        sys.exit(1)
+
+
+def restart_telemetry_server():
+    """Restarts the telemetry server process by Supervisord and then checks
+    it is actually running.
+
+    Args:
+        None
+
+    Returns:
+        None
+    """
+    restart_telemetry_server_cmd = "supervisorctl restart telemetry"
+    retry_times = 0
+
+    while retry_times <= MAX_RETRY_TIMES:
+        retry_times += 1
+        exit_code, command_stdout = get_command_result(restart_telemetry_server_cmd)
+        if exit_code != 0:
+            syslog.syslog(syslog.LOG_INFO,
+                          "Failed to restart telemetry server process and retry after 60 seconds ...")
+            time.sleep(60)
+        else:
+            break
+
+    if retry_times > MAX_RETRY_TIMES:
+        syslog.syslog(syslog.LOG_ERR,
+                      "Failed to restart telemetry server process after trying '{}' times and exiting ..."
+                      .format(MAX_RETRY_TIMES))
+        sys.exit(2)
+
+    check_telemetry_server_running()
+
+
+def check_certificate_rotated(certificate_file_name):
+    """Leverages the 'inotify' module to monitor the file system events under the
+    directory which stores the SONiC credentials and restarts telemetry server
+    process if its certificate was rotated.
+
+
+    Args:
+        certificate_file_name: A string indicates the telemetry certificate file name.
+
+    Returns:
+        None.
+    """
+    certificate_file_rotated = False
+
+    inotify_instance = inotify.adapters.Inotify()
+    inotify_instance.add_watch(CREDENTIALS_DIR_PATH)
+    for event in inotify_instance.event_gen(yield_nones=False):
```

Comment:
>inotify_instance [](http://example.com/codeflow?start=17&length=16)

In extreme case, the file is deleted by a malicious user, will the inotify_instance still working? I think its link to inode, and deleting file will destroy the inode.

If this is true, a crash is better than a dead loop.

Reply:
The `inotify` is `inode` based and it will monitor the credentials directory `/etc/sonic/credentials/` to see whether the telemetry certificate file was rotated or not. If certificate file was deleted by accidentally, the `inotify_instance` will not be impacted.

I updated the PR to log an error message if the certificate was deleted. What I am thinking is if the certificate was restored later, then it can be treated as a kind of `rotation` operation and the telemetry server will be restarted by this script.